Carousel: misc improvements

yomed opened the pull request:
## Description
- avoid use of `update_` methods, so all state goes through marko
- clean up current `FIXME`s in state handling
- put DOM manipulation inside `requestAnimationFrame`
- normalize `index` values when out of bounds
- avoid processing for carousels that don't have room to move at all
- clarify docs on `touch` usage

## References
Fixes #161 
Fixes #177

ianmcburnie commented:
Also fixes this issue I found on demo page:

![screen shot 2018-05-15 at 11 57 04 am](https://user-images.githubusercontent.com/38065/40077849-b39685d4-5837-11e8-8aff-b9ca1e79bf0d.png)

yomed commented:
FYI I'm working on a larger refactor to fix https://github.com/eBay/ebayui-core/issues/177, but that should probably be kept separate from this safer set of changes.

yomed commented:
@DylanPiercey I've updated this PR with the latest. The diff isn't too helpful, but basically I tried to cover all the fundamental state issues we talked about. I also added a new test to cover the parent state case.

DylanPiercey commented on src/components/ebay-carousel/index.js:
```diff
@@ -85,120 +85,104 @@ function init() {
-            this.setState('index', this.calculatePrevIndex());
-        }
-        emitAndFire(this, 'carousel-prev', { originalEvent });
+    if (this.state.index > this.lastIndex) {
```
Shouldn't this be `this.state.index !== this.lastIndex`? I might be missing something.

yomed replied:
`lastIndex` is used as in "A carousel of length 4 would have lastIndex=3." I'm using `activeIndex` below to keep track of when to handle the state updates.

DylanPiercey commented on src/components/ebay-carousel/index.js:
```diff
@@ -85,120 +85,104 @@ function init() {
-        const oldFirstVisibleIndex = this.state.firstVisibleIndex;
-        const oldLastVisibleIndex = this.state.lastVisibleIndex;
+function processMovement() {
+    this.processMovementFrame = requestAnimationFrame(() => {
```
We should probably `cancelAnimationFrame` before this line. Just incase they change the state twice in a frame we can avoid doing extra work.
